service/src/test/java/bio/terra/cbas/initialization/TestCloneRecoveryService.java

```diff
@@ -0,0 +1,291 @@
+package bio.terra.cbas.initialization;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.when;
+
+import bio.terra.cbas.config.CbasContextConfiguration;
+import bio.terra.cbas.dao.MethodDao;
+import bio.terra.cbas.dao.MethodVersionDao;
+import bio.terra.cbas.dao.RunDao;
+import bio.terra.cbas.dao.RunSetDao;
+import bio.terra.cbas.dao.util.ContainerizedDaoTest;
+import bio.terra.cbas.models.*;
+import java.time.OffsetDateTime;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.*;
+import org.mockito.Mock;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class TestCloneRecoveryService extends ContainerizedDaoTest {
```

Comment:
Feel free to push back on this, kind of an opinionated statement:

In an ideal world, I think a Service test (which validates business logic) would not need test containers. It would be 75% unit tests checking in-memory data transformations are calculated properly and 25% mocked-dao tests validating that the glue was calling the right functions on the dao. Then the test containers would only be needed for the very final "did our database connection commands get implemented correctly in SQL queries"

Reply:
I've developed some thoughts as I actually work on this - maybe we can find a time to talk later today

Reply:
I'll write them down to clarify my thoughts, but let's still chat...

- I agree with your philosophy, 100%
- I think the mistake here is calling this a "service"
- I wrote the tests this way because it gave me confidence that it was working as intended. As I try to rewrite, I feel less good about the tests that emerge
- Each of the three helper methods that `pruneCloneSourceWorkspaceHistory` uses feel like they could be DAO methods - I just happened to have written them with Java and existing DAO methods rather than new SQL

So I think what might make the most sense is to move the helpers to their respective DAO classes, and do some basic unit testing (with mocks) of `pruneCloneSourceWorkspaceHistory`.

Reply:
Just merged in a substantial refactor to improve testability